### PR TITLE
[opentitantool]: Allow SPI chip select across transactions

### PR DESCRIPTION
**Pull request by jesultra**
While implementing TPM communication, I came across a use case in which
I wanted to perform a few SPI transfers, inspect the data that was read,
and then possibly issue some more SPI transfers, while keeping CS
asserted throughout.

The current `spi::Target` interface is such that `run_transaction()`
keeps CS asserted for the duration of the list of transfers given in the
parameter list and then deasserts it unconditionally, with no option to
add more based on the result of the previous ones.

This PR introduces an explicit `assert_cs()` function, which can be used
to keep CS asserted across multiple invocations of `run_transaction()`,
like below.

```
let spi = ...;
{
  let _cs_asserted = spi.assert_cs()?;
  spi.run_transaction(...)?;
  if ... {
    spi.run_transaction(...)?;
  }
  // CS will automatically deassert here (or earlier if bailing out.)
}
```

Furthermore, `assert_cs()` counts its invocations, and releases CS only
when every nested assertion has been released, such that if the above
code was part of a helper method, and the called wanted to make several
invocations of the helper, during a single CS session, then it could put
another `assert_cs()` around its logic.

Change-Id: I219f3647d29129c4dfc1fc346f855fb4e9da2e53

### CODE REVIEW

File: sw/host/opentitanlib/src/io/spi.rs [hunk 1]
```diff
@@ -151,6 +151,39 @@ pub trait Target {
...
-    /// Runs a SPI transaction composed from the slice of [`Transfer`] objects.
+    /// Runs a SPI transaction composed from the slice of [`Transfer`] objects.  Will assert the
+    /// CS for the duration of the entire transactions.
     fn run_transaction(&self, transaction: &mut [Transfer]) -> Result<()>;
+
+    /// Assert the CS signal.  Uses reference counting, will be deasserted when each and every
+    /// returned `AssertChipSelect` object have gone out of scope.
+    fn assert_cs<'a>(&'a self) -> Result<AssertChipSelect<'a>> {
+        Err(SpiError::InvalidOption("This target does not support assert_cs".to_string()).into())
+    }
```
Comment (cfrantz) [inline sw/host/opentitanlib/src/io/spi.rs:162]: How is this going to interact with the network proxy transport?

Comment (jesultra) [reply sw/host/opentitanlib/src/io/spi.rs:162]: We will need new protocol requests to assert and deassert the CS (increment/decrement the count, really).  This method will be implemented in the client library by sending an increment request, and returning an object which when dropped sends the decrement request.  On the network server end, we probably are going to have a vector of AssertChipSelect objects associated with each connection, handling the increment request by invoking assert_cs on the underlying transport and putting its result on the vector, and handling decrement requests by removing an element and dropping it.  (If the connection is broken, the stack of objects will also be dropped, releasing any chip select on the underlying transport.)

In short, I think it can be done.  (One could avoid the vector in favor of an Option<AssertChipSelect> and a separate integer, to have the proxy server do the counting.)

**Comment by timothytrippel** [comment]
are we ok to merge this (after CI passes?)

@jesultra, you will have to rebase this on master and repush as we had some CI issues earlier this week.

**Comment by jesultra** [comment]
I do not want to merge as is, since only a single transport has implementation.  What I was hoping to get is informal approval that this approach makes sense, then I would implement the interface for the other transports, possibly including the proxy transport, before formally asking for it to be merged.

File: sw/host/opentitanlib/src/transport/ultradebug/spi.rs [hunk 2]
```diff
@@ -42,9 +46,34 @@ impl UltradebugSpi {
...
+        let device = self.device.borrow();
+        // Assert CS# (drive low).
+        device
+            .execute(&mut [mpsse::Command::SetLowGpio(
+                device.gpio_direction,
+                device.gpio_value & !Self::MASK_CHIP_SELECT,
+            )])
+            .context("FTDI error")?;
+        Ok(())
+    }
```
Comment (cfrantz) [inline sw/host/opentitanlib/src/transport/ultradebug/spi.rs:64]: Would this function be better as:
```
fn assert_cs(&mut self, assert: bool) -> Result<()>
```

Where the function does handles the reference counting and command construction based on the value of `assert`?

Comment (jesultra) [reply sw/host/opentitanlib/src/transport/ultradebug/spi.rs:64]: I suppose that I could save some lines by consolidating the two internal methods into one.  I will do so, as I implement this API on the other transport backends.

File: sw/host/opentitanlib/src/transport/ultradebug/spi.rs [hunk 3]
```diff
@@ -137,12 +168,35 @@ impl Target for UltradebugSpi {
...
+}
+
+impl TargetChipDeassert for UltradebugSpi {
+    fn deassert_cs(&self) {
+        let mut inner = self.inner.borrow_mut();
+        inner.cs_asserted_count -= 1;
+        if inner.cs_asserted_count == 0 {
+            // We cannot propagate errors through `Drop::drop()`, so panic on any error.  (Logging
+            // would be another option.)
+            self.do_deassert_cs().unwrap();
```
[the remark this answers is not in the document]
Comment (jesultra) [reply sw/host/opentitanlib/src/transport/ultradebug/spi.rs:199]: Good point, I will do so.

File: sw/host/opentitanlib/src/io/spi.rs (hunk 1 above)
Comment (jesultra) [reply sw/host/opentitanlib/src/io/spi.rs:162]: My problem now is that in TransportCommandHandler.do_execute_cmd() we have a &'a TransportWrapper on which a call to spi() is made, returning a Rc<dyn spi::Target>.  We want to invoke the new assert_cs() on the spi::Target borrowed from that Rc, but for all the compiler knows all the Rc objects could run out of scope before the AssertChipSelect, if we had put in on a vector.
In order to switch to runtime lifetime management, the AssertChipSelect would probably need to hold a Rc<dyn spi::Target>, rather than a direct reference to the object.  This could possibly be done by modifying the signature into:
```
fn assert_cs(self: Rc<dyn Target>) -> Result<AssertChipSelect>;
```
I am not sure if it is possible to create methods this way, though.

Comment (cfrantz) [reply sw/host/opentitanlib/src/io/spi.rs:162]: There are a few forms of `self` allowed.  Basically, any "smart pointer" formulation of `self` is permitted:
```
fn assert_cs(self: &Rc<Self>) -> Result<foo>
```

Comment (jesultra) [reply sw/host/opentitanlib/src/io/spi.rs:162]: I will try this.  However, my original idea was to not have the deassert() method be part of the spi::Target trait, and instead have the common AssertChipSelect struct invoke deasser() through a second trait.  Any struct implementing spi::Target would also implment my second "hidden" trait, TargetChipDeassert, and that way be able to construct a AssertChipSelect passing its self cast into a dyn TargetChipDeassert.  When I instead have a Rc<dyn Target> or Rc of some specific implementing struct, I do not think that I can similarly upcast to Rc<dyn TargetChipDeassert>.

I will see what I can do, but I am afraid that I will not be able to find a way for AssertChipSelect to hold a Rc<> reference to one out of various implementation spi::Target, in such a way that it can invoke a method that is not exposed as part of the trait.

**Comment by jesultra** [comment]
Now this PR has implementation of the SPI chip select interface for UltraDebug, HyperDebug, and for the proxy client/server implementation.  I consider it to be demonstration enough that I want to merge it in this state.  If the need arises, we can later implement the new method for one or more of the remaining transport backends as well, provided that the hardware supports it.

File: sw/host/opentitanlib/src/io/spi.rs (hunk 1 above)
Comment (jesultra) [reply sw/host/opentitanlib/src/io/spi.rs]: The only thorn is that Rust apparently does not support `self: &Rc<Self>`, but only supports the non-reference version, in order to be "object safe".  The description in the link gives me the impression that it is not an inherent limitation of the type system, but rather a matter of it not being implement, yet.  We can seamlessly remedy this in the future, if the Rust compile should get support, for now, the caller will have to clone the `Rc` ahead of calling `assert_cs()`.  I think that is acceptable.  
https://doc.rust-lang.org/reference/items/traits.html#object-safety

File: sw/host/opentitanlib/src/transport/hyperdebug/spi.rs [hunk 4]
```diff
@@ -346,4 +405,28 @@ impl Target for HyperdebugSpiTarget {
...
+}
+
+impl TargetChipDeassert for HyperdebugSpiTarget {
+    fn deassert_cs(&self) {
+        let cs_asserted_count = self.cs_asserted_count.get();
+        if cs_asserted_count - 1 == 0 {
+            // We cannot propagate errors through `Drop::drop()`, so panic on any error.  (Logging
+            // would be another option.)
+            self.do_assert_cs(false)
+                .expect("Error while deasserting CS");
```
Comment (cfrantz) [inline sw/host/opentitanlib/src/transport/hyperdebug/spi.rs:428]: How will this panic affect the opentitan_session program?  Will this cause program termination?

Depending on the severity of this error (ie: requires manual intervention to fix), panic may be the right action.  However, if destroying and restarting a session works, then `log::error!` might be better.

Ok to leave this as-is for now and figure it out later.

File: sw/host/opentitanlib/src/transport/hyperdebug/spi.rs [hunk 5]
```diff
@@ -346,4 +405,28 @@ impl Target for HyperdebugSpiTarget {
         }
         Ok(())
     }
+
+    fn assert_cs<'a>(self: Rc<Self>) -> Result<AssertChipSelect> {
```
[the remark this answers is not in the document]
Comment (jesultra) [reply sw/host/opentitanlib/src/transport/hyperdebug/spi.rs:409]: Right, odd that the Rust compiler does not warn that the lifetime parameter is unused, (or that it is not present in the method signature of the trait.)

File: sw/host/opentitanlib/src/io/spi.rs [hunk 6]
```diff
@@ -151,6 +151,37 @@ pub trait Target {
...
+/// Object that keeps the CS asserted, deasserting when it goes out of scope, (unless another
+/// instance keeps CS asserted longer.)
+pub struct AssertChipSelect {
+    target: Rc<dyn TargetChipDeassert>,
+}
+
+impl AssertChipSelect {
+    // Needs to be public in order for implementation of `Target` to be able to call it.  Never
+    // called by users of `Target`.
+    pub fn create(target: Rc<dyn TargetChipDeassert>) -> Self {
```
[the remark this answers is not in the document]
Comment (jesultra) [reply sw/host/opentitanlib/src/io/spi.rs:172]: Done

File: sw/host/opentitanlib/src/proxy/handler.rs [hunk 7]
```diff
@@ -175,6 +180,25 @@ impl<'a> TransportCommandHandler<'a> {
...
+                            .or_insert(Vec::new())
+                            .push(instance.assert_cs()?);
+                        Ok(Response::Spi(SpiResponse::AssertChipSelect))
+                    }
+                    SpiRequest::DeassertChipSelect => {
+                        // Remove a `spi::AssertChipSelect` object from the stack for this
+                        // particular SPI instance.
+                        self.spi_chip_select
+                            .get_mut(id)
+                            .ok_or(anyhow::anyhow!("Unbalanced CS"))?
```
[the remark this answers is not in the document]
Comment (jesultra) [reply sw/host/opentitanlib/src/proxy/handler.rs:197]: Done

File: sw/host/opentitanlib/src/transport/proxy/spi.rs [hunk 8]
```diff
@@ -143,4 +145,23 @@ impl Target for ProxySpi {
...
+}
+
+impl TargetChipDeassert for ProxySpi {
+    fn deassert_cs(&self) {
+        match self
+            .execute_command(SpiRequest::DeassertChipSelect)
+            .expect("Error deactivating chip select")
+        {
+            SpiResponse::DeassertChipSelect => (),
+            _ => panic!("Error deactivating chip select"),
```
[the remark this answers is not in the document]
Comment (jesultra) [reply sw/host/opentitanlib/src/transport/proxy/spi.rs:164]: It is correct that this is client code.  The problem is that this method has no means of returning an `Err()` as it is called from within a `drop()` method.
As such, the only option is to `panic!`, or to log something and continue.  However, if there is a communication problem with the server, as we are trying to release chip select, I believe that logging and continuing could lead to unexpected behavior, as further execution could go on, thinking that it repeatedly activates and deactivates chip select, when in reality, it is continuously asserted, due to the one request to release not being received by the session process, meaning that the count never again reaches zero.  This could potentially lead to the SPI target device mis-interpreting the data bytes sent to it, and performing operations that was not intended.  (E.g. if a write command first takes the address, and then any number of data bytes to be written on consecutive addresses, then missing a chip select deassertion could lead to the next commands being treated as further data bytes, overwriting areas that were not intended to be written to.)

**Comment by jesultra** [comment]
I believe that I have addressed all comments, please let me know if additional modifications are needed before this can be merged.
